code/krepel/string/utf_conversion.d

```diff
@@ -0,0 +1,607 @@
+      auto CodePoint = ExtractCodePoint(Data, Endiannes);
+      Data = Data[Size..$];
+      auto UTF8 = UTF8FromCodePoint(CodePoint);
+      Result.Concat(UTF8[0..$]);
+    }
+    // Skip not parseable bytes
+    else
+    {
+      Data = Data[1..$];
+    }
+  }
+
+  return Result;
+}
+
+WString ToUTF16(UString String, Endian Endiannes = Endian.Little)
+{
+  WString Result = WString(String.Allocator);
+  auto Data = String[][];
+  while(Data.length)
+  {
+    auto Size = CharSize(Data);
+    if (Size)
+    {
+      auto CodePoint = ExtractCodePoint(Data);
+      Data = Data[Size..$];
+      auto UTF16 = UTF16FromCodePoint(CodePoint, Endiannes);
+      Result.Concat(UTF16[0..$]);
+    }
+    // Skip not parseable bytes
+    else
+    {
+      Data = Data[1..$];
+    }
+  }
+
+  return Result;
+}
+
+unittest
+{
+  char Test = 'A';
+  uint CodePoint = ExtractCodePoint([Test]);
+  auto Result = UTF8FromCodePoint(CodePoint);
+
+  assert(Result.length == 1);
+  assert(Test == Result[0]);
+
+  auto Test2 = "ü"c;
+  CodePoint = ExtractCodePoint(Test2);
+  Result = UTF8FromCodePoint(CodePoint);
+
+  assert(Test2.length == Result.length);
+  assert(Test2[0] == Result[0]);
+  assert(Test2[1] == Result[1]);
+
+  auto Test3 = "𐐷"c;
+  CodePoint = ExtractCodePoint(Test3);
+  assert(CodePoint == 0x10437U);
+  Result = UTF8FromCodePoint(CodePoint);
+
+  assert(Test3.length == Result.length);
+  assert(Test3.length == 4);
+  assert(Test3[0] == Result[0]);
+  assert(Test3[1] == Result[1]);
+  assert(Test3[2] == Result[2]);
+  assert(Test3[3] == Result[3]);
+
+}
+
+unittest
+{
+  wchar Test = 'A';
+  uint CodePoint = ExtractCodePoint([Test], Endian.Little);
+  auto Result = UTF16FromCodePoint(CodePoint, Endian.Little);
+
+  assert(Result.length == 1);
+  assert(Test == Result[0]);
+
+  auto Test2 = "𤭢"w;
+  auto Test3 = "𤭢"c;
+  auto UTF8CodePoint = ExtractCodePoint(Test3);
+
+  CodePoint = ExtractCodePoint(Test2, Endian.Little);
+  Result = UTF16FromCodePoint(CodePoint, Endian.Little);
+  assert(CodePoint == 0x24B62);
+  assert(CodePoint == UTF8CodePoint);
+  assert(Test2.length == Result.length);
+  assert(Result.length == 2);
+  assert(Test2[0] == Result[0]);
+  assert(Test2[1] == Result[1]);
+
+}
+
+unittest
+{
+  import krepel.memory;
+  auto Allocator = CreateTestAllocator();
+
+  auto UTF16String = WString("TestString", Allocator);
+  auto UTF8String = UString("TestString", Allocator);
+  auto ConvertedString = UTF16String.ToUTF8();
+  assert(ConvertedString == UTF8String);
+}
+
+unittest
+{
+  import krepel.memory;
+  auto Allocator = CreateTestAllocator();
+
+  auto UTF16String = WString("TestString", Allocator);
+  auto UTF8String = UString("TestString", Allocator);
+  auto ConvertedString = UTF8String.ToUTF16();
+  assert(ConvertedString == UTF16String);
+}
+
+unittest
+{
+  import krepel.memory;
+  auto Allocator = CreateTestAllocator();
+
+  auto String = UString("Tääst", Allocator);
+  assert(String.CharCount == 5);
+  assert(String.Count == 7);
```

Comment:
I know. But the way I see it, a string is a container of characters rather than a container of bytes, so personally I would expect a property called `Count` to return the number of string characters rather than the number of bytes stored. Since I can ask a string for 2 different kinds of counts, I think we should disambiguate the names here.

In fact, I'd be ok with `Count` returning the character count and using `Data.Count` if I need the number of bytes. The `Data` member is public anyway, so that already works.